docs/specs/saml.md

```diff
@@ -0,0 +1,697 @@
+  - Digest Method ([DigestMethod](https://www.w3.org/TR/xmldsig-core1/#sec-DigestMethod)) is configurable with config `saml.signing.digest_method`. Please see below for supported values.
+- If Service Providers has signing certs uploaded, all requests and responses from the Service Provider will be validated against the provided certs.
+  - To facilitate cert rotations, two certs are allowed to be uploaded. When there are two certs, the requests and responses will be considered valid if its signature can be validated by any one of the uploaded certs.
+
+### <a id="6_1"></a> Configs
+
+```yaml
+saml:
+  signing:
+    key_id: key01
+    signature_method: RSAwithSHA256
+    digest_method: SHA256
+```
+
+- `saml.signing`: Required. Configurations of signing behvaiors. MUST be an object with the following fields:
+  - `key_id`: Required. A string which points to a signing key secret inside the `saml.idp.signing` secret. See the below [Secrets] section for details.
+  - `signature_method`: Optional. An enum which specifies the signing algorithm used to generate the signature. The only supported value is `RSAwithSHA256`. If not provided, it defaults to `RSAwithSHA256`. For details, see [xmldsig](https://www.w3.org/TR/xmldsig-core1/#sec-SignatureAlg).
+  - `digest_method`: Optional. An enum which specifies the digest method used. The supported values are `SHA1` and `SHA256`. If not provided, it defaults to `SHA256`. For details, see [xmldsig](https://www.w3.org/TR/xmldsig-core1/#sec-MessageDigests).
+
+### <a id="6_2"></a> Secrets
+
+Two secrets are defined for SAML signing:
+
+```yaml
+- data:
+    certs:
+      - cert:
+          pem: "-----BEGIN CERTIFICATE-----\nMIIH/TCCBeWgAwIBAgIQaBYE3/..."
+        key:
+          alg: RS256
+          created_at: 1671439710
+          kid: key01
+          d: xxxxxx
+          use: sig
+      - cert:
+          pem: "-----BEGIN CERTIFICATE-----\nMIIH/TCCBeWgAwIBAgIQaBYE3/..."
+        key:
+          alg: RS256
+          created_at: 1671439710
+          kid: key02
+          d: xxxxxx
+          use: sig
+  key: saml.idp.signing
+```
+
+- The above secret is for storing the signing key pairs used by Authgear to sign the SAML requests, responses and assertions. The keys are stored in jwk format.
+  - `key`: Must be `saml.idp.signing`.
+  - `data`: Required. An object containing the following fields:
+    - `certs`: Required. A list of objects which represents the signing keypairs and the X.509 certs used to sign the IdP requests, responses and assertions. The minimum number of items in the list is 1. The maximum number of items in the list is 0. The item inside the list MUST contains the following fields:
+      - `cert`: Requried. An object represents a X.509 cert. It MUST contains the following fields:
+        - `pem`: Required. The X.509 cert in pem format. See [Metadata](#7) for details.
```

Comment:
What fields are required in the X.509 certificate? I wonder if we can just store the private key JWK, and generate the X.509 certificate on the fly?

Reply:
I think it need to at least have fields required by a X.509 cert, such as "Not Before", "Not After"...
> I wonder if we can just store the private key JWK, and generate the X.509 certificate on the fly?

If we always generate a new cert, the timestamps mentioned above will always be changing.

Reply:
> I think it need to at least have fields required by a X.509 cert, such as "Not Before", "Not After"...

We can use the JWT claim `exp` and `iat`, `nbf` in the JWK for this. https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4 Then we should be able to generate a X.509 certificate that is stable and not changing.

Reply:
I realized we have been using the claim `created_at` to store the creation time of a JWK.

Reply:
However, it seems jwk are not allowed to have such fields?
https://datatracker.ietf.org/doc/html/rfc7517#section-4
https://datatracker.ietf.org/doc/html/rfc7518#section-6.3.2

Reply:
And actually, storing the exact cert also has a benefit that this avoid us to change the the cert accidentally (Such as when we supported a new algo in future, or changing the subject name of the cert for some reason, we don't want to affect the old certs).